### DISCUSSION TITLE
updateTriggers not working for color fill in a GeoJSON layer

### DISCUSSION THREAD

**Discussion by arthurgailes** [visgl/deck.gl#6197] [Q&A] [answered]
Hello,

I'm attempting to update the polygon fill on a GeoJSON layer where I update the data property of the layer. The code looks like this:
```

const render = (deckgl, shp) => {
// update the data here
 shp.features.forEach(function test(d) {
    const val =myFunc(); // update the val column
    Object.assign(d.properties, { value: val || 0 });
    return d;
  });

  const setting = {
    id: 'block_poly',
    data: shp,
    getFillColor(d) {
      debugger;
      return color_scale(Math.random())
    },
    updateTriggers: {
      data: Math.random(), // tried turning this on and off
      getFillColor: Math.random(),
    },
  });

  let { layers } = deckgl.props;

  const layer = new GeoJsonLayer(setting);

  // either add or update the layer
  const idx = layers.findIndex((l) => l.props.id === layer.id);

  if (idx >= 0) layers[idx] = layer;
  else layers = layers.concat(layer);

  deckgl.setProps({ layers });
}
```

When I update the layer, the updateTrigger value in `props` shows that it's been updated to the new number, but the getFillColor only runs on the first load (verified with debugger). I'm also pretty sure it "knows" it's been updated, because the count variable iterates properly. What am I missing?

**Comment by Pessimistress** [comment, marked as answer]
- `updateTriggers` is for accessor functions (`get*`). `updateTriggers.data` has no effect.
- you are calling `deckgl.setProps({layers})` with the same array. From deck's perspective nothing has changed. Don't mutate the existing layers array, make a new one.

**Reply by arthurgailes** [reply]
Thanks very much!  I resolved it by cloning; is this an efficient method to handle such updates for many (~20) layers? The optimization section in the docs doesn't mention it.

```
let layers = deckgl.props.layers.map((l) => l.clone());
layers[idx] = new GeoJsonLayer({settings})
```

**Reply by Pessimistress** [reply]
You don't need to clone layers that are not changed. You just need `layers` to be a different object, e.g. `layers = [...layers]`

**Reply by arthurgailes** [reply]
Thanks so much for your quick help. I wouldn't have made my deadline without you!
